vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs

```diff
@@ -74,7 +82,45 @@ type internal SingleFileWorkspaceMap(workspace: VisualStudioWorkspace,
 
         member _.OnAfterFirstDocumentLock(_, _, _, _) = VSConstants.E_NOTIMPL
 
-        member _.OnAfterSave(_) = VSConstants.E_NOTIMPL
+        member _.OnAfterSave(docCookie) = 
+            // This causes re-analysis to happen when a F# document is saved.
+            // We do this because FCS relies on the file system and existing open documents
+            // need to be re-analyzed so the changes are propogated.
+            // We only re-analyze F# documents that are dependent on the document that was just saved.
+            // We ignore F# script documents here.
+            let vsTextBuffer = rdt4.GetDocumentData(docCookie) :?> IVsTextBuffer
+            let textBuffer = editorAdaptersFactory.GetDataBuffer(vsTextBuffer)
+
+            let textContainer = textBuffer.AsTextContainer()
+            let mutable workspace = Unchecked.defaultof<_>
+            if Workspace.TryGetWorkspace(textContainer, &workspace) then
+                let solution = workspace.CurrentSolution
+                let documentId = workspace.GetDocumentIdInCurrentContext(textContainer)
+                match box documentId with
+                | null -> ()
+                | _ -> 
+                    let document = solution.GetDocument(documentId)
+                    if document.Project.Language = LanguageNames.FSharp && not document.IsFSharpScript then
+                        let openDocIds = workspace.GetOpenDocumentIds()
+                        let depProjIds = document.Project.GetDependentProjectIds().Add(document.Project.Id)
+
+                        let docIdsToReanalyze =
+                            openDocIds
+                            |> Seq.filter (fun x ->
+                                depProjIds.Contains(x.ProjectId) && x <> document.Id &&
+                                (
+                                    let doc = solution.GetDocument(x)
+                                    match box doc with
+                                    | null -> false
+                                    | _ -> doc.Project.Language = LanguageNames.FSharp
+                                )
+                            )
+                            |> Array.ofSeq
+                        if Array.isEmpty docIdsToReanalyze then ()
+                        else
+                            analyzerService.Reanalyze(workspace, documentIds=docIdsToReanalyze)
```

Comment:
What priority is given to this reanalysis internally in Roslyn?  I'm concerned that it's given a high-priority, when we really want it to be low-priority except for the in-focus file (and if a file becomes in-focus then it is promoted to high-priority).  I know Roslyn has a priority queue of some kind, I'm not quite sure how this works though.

Reply:
I mean, it's likely the same as what Roslyn is already scheduling every time you edit a file? So I guess it's ok.  

DO you know if this code run on the foreground UI thread of background thread?

Reply:
By default, it's not high-priority. After manually testing this, calling the Reanalyze doesn't immediately analyze the given documents on call; it does wait a second or two.

Reply:
My guess is that the code is the foreground; but I will double check. It might be ok considering Reanalyze really just queues up work.

Reply:
This shouldn't be an issue as I can do a fire and forget call